internal/checker/services.go

```diff
@@ -0,0 +1,426 @@
+// computing whether this is a `super` property access.
+// type: the type whose property we are checking.
+// property: the accessed property's symbol.
+func (c *Checker) IsValidPropertyAccessForCompletions(node *ast.Node, t *Type, property *ast.Symbol) bool {
+	return c.isPropertyAccessible(
+		node,
+		node.Kind == ast.KindPropertyAccessExpression && node.Expression().Kind == ast.KindSuperKeyword,
+		false, /*isWrite*/
+		t,
+		property,
+	)
+	// Previously we validated the 'this' type of methods but this adversely affected performance. See #31377 for more context.
+}
+
+func (c *Checker) GetAllPossiblePropertiesOfTypes(types []*Type) []*ast.Symbol {
+	unionType := c.getUnionType(types)
+	if unionType.flags&TypeFlagsUnion == 0 {
+		return c.getAugmentedPropertiesOfType(unionType)
+	}
+
+	props := createSymbolTable(nil)
+	for _, memberType := range types {
+		augmentedProps := c.getAugmentedPropertiesOfType(memberType)
+		for _, p := range augmentedProps {
+			if _, ok := props[p.Name]; !ok {
+				prop := c.createUnionOrIntersectionProperty(unionType, p.Name, false /*skipObjectFunctionPropertyAugment*/)
+				// May be undefined if the property is private
+				if prop != nil {
+					props[p.Name] = prop
+				}
+			}
+		}
+	}
+	return slices.Collect(maps.Values(props))
+}
+
+func (c *Checker) IsUnknownSymbol(symbol *ast.Symbol) bool {
+	return symbol == c.unknownSymbol
+}
+
+func (c *Checker) IsUndefinedSymbol(symbol *ast.Symbol) bool {
+	return symbol == c.undefinedSymbol
+}
+
+func (c *Checker) IsArgumentsSymbol(symbol *ast.Symbol) bool {
+	return symbol == c.argumentsSymbol
+}
+
+// Originally from services.ts
+func (c *Checker) GetNonOptionalType(t *Type) *Type {
+	return c.removeOptionalTypeMarker(t)
+}
+
+func (c *Checker) GetStringIndexType(t *Type) *Type {
+	return c.getIndexTypeOfType(t, c.stringType)
+}
+
+func (c *Checker) GetNumberIndexType(t *Type) *Type {
+	return c.getIndexTypeOfType(t, c.numberType)
+}
+
+func (c *Checker) GetCallSignatures(t *Type) []*Signature {
+	return c.getSignaturesOfType(t, SignatureKindCall)
+}
+
+func (c *Checker) GetConstructSignatures(t *Type) []*Signature {
+	return c.getSignaturesOfType(t, SignatureKindConstruct)
+}
+
+func (c *Checker) GetApparentProperties(t *Type) []*ast.Symbol {
+	return c.getAugmentedPropertiesOfType(t)
+}
+
+func (c *Checker) getAugmentedPropertiesOfType(t *Type) []*ast.Symbol {
+	t = c.getApparentType(t)
+	propsByName := createSymbolTable(c.getPropertiesOfType(t))
+	var functionType *Type
+	if len(c.getSignaturesOfType(t, SignatureKindCall)) > 0 {
+		functionType = c.globalCallableFunctionType
+	} else if len(c.getSignaturesOfType(t, SignatureKindConstruct)) > 0 {
+		functionType = c.globalNewableFunctionType
+	}
+
+	if functionType != nil {
+		for _, p := range c.getPropertiesOfType(functionType) {
+			if _, ok := propsByName[p.Name]; !ok {
+				propsByName[p.Name] = p
+			}
+		}
+	}
+	return c.getNamedMembers(propsByName)
+}
+
+func IsUnion(t *Type) bool {
```

Comment:
I think these could all be methods on Type, right? Or is this just for easier porting?